src/Language/PureScript/Make/Actions.hs

```diff
@@ -279,18 +313,41 @@ buildMakeActions outputDir filePathMap foreigns usePrefix =
   writeCacheDb :: CacheDb -> Make ()
   writeCacheDb = writeCacheDb' outputDir
 
+  writePackageJson :: Make ()
+  writePackageJson = writePackageJson' outputDir
+
+data ForeignModuleType = ESModule | CJSModule deriving (Show)
+
 -- | Check that the declarations in a given PureScript module match with those
 -- in its corresponding foreign module.
-checkForeignDecls :: CF.Module ann -> FilePath -> Make ()
+checkForeignDecls :: CF.Module ann -> FilePath -> Make (ForeignModuleType, S.Set Ident)
 checkForeignDecls m path = do
   jsStr <- T.unpack <$> readTextFile path
-  js <- either (errorParsingModule . Bundle.UnableToParseModule) pure $ JS.parse jsStr path
+  js <- either (errorParsingModule . Bundle.UnableToParseModule) pure $ JS.parseModule jsStr path
+
+  (foreignModuleType, foreignIdentsStrs) <-
+    case (,) <$> getForeignModuleExports js <*> getForeignModuleImports js of
+      Left reason -> errorParsingModule reason
+      Right (Bundle.ForeignModuleExports{..}, Bundle.ForeignModuleImports{..})
+        | not (null cjsExports && null cjsImports)
+        , null esExports
+        , null esImports -> do
+            let deprecatedFFI = filter (elem '\'') cjsExports
+            unless (null deprecatedFFI) $
+              errorDeprecatedForeignPrimes deprecatedFFI
+
+            when (elem "default" cjsExports) $
+              errorDeprecatedFFIDefaultCJSExport
```

Comment:
If we want to go with this approach I’d like to emit this error as a warning starting from v0.14.1.

Reply:
That is, I think a spec-conformant implementation of ES modules will treat `export default foo` identically to `exports.default = foo`.

Reply:
Ah, ok, so what I was missing was that a) `default` isn't allowed as a variable name, and b) in node.js, the `exports` object is treated as the default export when importing CJS from ESM.
```javascript
// test.js
import thing1 from "./def.cjs"
import thing2 from "./def.js"

console.log(thing1);
console.log(thing2);

// def.cjs
exports.default = "hello, world";

// def.js
export default "hello, world";
```
produces
```
$ node test.js
{ default: 'hello, world' }
hello, world
```
so if we are going to be using named exports in ESM, then yeah, I can't think of another way around it. Warning on CJS "default" exports also seems reasonable to me given that we are going to drop support for CJS FFI modules eventually anyway.

Reply:
If the ES wrapper imports the default export of the CommonJS foreign module and re-export its properties with export declarations as I thought about in https://github.com/purescript/purescript/pull/3791#discussion_r611029477 we could keep supporting CommonJS exports named default:

```js
import $foreign from "./foreign.cjs";

var $$default = $foreign.default;
export { $$default as default };
```